### PR TITLE
Use systemd-run to spawn exec/shell sessions

### PR DESCRIPTION
**Pull request by lachypjones**
# Description

# Self-review quick check

 * [ ] Make decisions that cost a lot to reverse explicit in the PR description.
 * [ ] Avoid nested conditions.
 * [ ] Delete dead code and redundant comments.
 * [ ] Normalise symmetries by sticking to doing identical things identically. 
 ```
 // one way to handle errors
 if err := f(); err != nil {
    ...
 }
 
 // one way to handle multiple returns
 val, err := f()
 if err != nil {
    ...
 }
 ...
 ```
 * [ ] Check that coupled code elements, files, and directories are adjacent. For example, test data is stored as close as possible to a test.
 * [ ] Put variable declaration and initialisation together.
 * [ ] Divide large expressions into digestable and self-explanatory ones. Use multiple variables if required.
 * [ ] Put a blank line between two logically different chunks of code.

## Docs

* [ ] I have checked and added or updated relevant documentation.
* [ ] I have checked and added or updated relevant release notes.
* [ ] I have included the technical author in the review.

Or:

* [ ] I confirm the PR has no implications for documentation.

### CODE REVIEW

File: cmd/workshop/exec.go [hunk 1]
```diff
@@ -187,9 +166,15 @@ func (c *CmdExec) Run(cmd *cobra.Command, av []string) error {
 		interactive = false
 	} else {
 		interactive = stdinIsTerminal && stdoutIsTerminal
+		// If both --pty and --pipe are specified, systemd-run will infer from
```
Comment (lachypjones) [inline cmd/workshop/exec.go:169]: This is a dead comment, I'll remove it at a later time

File: tests/main/interface-mount/task.yaml [hunk 2]
```diff
@@ -1,11 +1,11 @@
 summary: Check mount interface scenarios
 restore: |
-  workshop_exec remove ws-mount || true
```
Comment (lachypjones) [inline tests/main/interface-mount/task.yaml:3]: utils.sh is not sourced by restore, meaning that this command doesn't remove the workshop

File: cmd/workshop/exec.go [hunk 3]
```diff
@@ -92,9 +93,8 @@ Notes:
 
 - To start a workshop before running a terminal session, use 'workshop start'
 
-- The subcommand is a shorthand for 'workshop exec';
-  it launches the login shell for 'workshop',
-  the default non-privileged user in a workshop
+- You can set the working directory, environment variables, user and group ID
+  for the spawned shell; reasonable defaults are provided
```
Comment (akcano) [inline cmd/workshop/exec.go:97]: I'd very much like to see the relevant defaults listed, not just mentioned here.

Comment (lachypjones) [reply cmd/workshop/exec.go:97]: This matches the documentation for exec. Happy to expand.

File: cmd/workshop/exec.go [hunk 4]
```diff
@@ -127,11 +127,17 @@ func (c *CmdShellAlias) Command() *cobra.Command {
...
+	cmd.Flags().StringArrayVar(&c.execCommand.Env, "env", []string{}, "Set an environment variable, e.g. 'FOO=bar'; if only the name is provided, the value is inherited from the CLI environment.")
+	cmd.Flags().IntVar(&c.execCommand.UserId, "uid", 1000, "Run as a specific workshop user")
+	cmd.Flags().IntVar(&c.execCommand.GroupId, "gid", 1000, "Run as a member of a specific workshop group")
+	c.execCommand.Interactive = true
 	return cmd
 }
 
 func (c *CmdShellAlias) Run(cmd *cobra.Command, av []string) error {
-	return c.execCommand.Run(cmd, []string{av[0], "su", "-l", "workshop"})
+	return c.execCommand.Run(cmd, []string{av[0], "/bin/bash", "-l"})
```
Comment (dmitry-lyfar) [inline cmd/workshop/exec.go:140]: This does not set:

```
XDG_RUNTIME_DIR=/run/user/1000
DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus
```

Comment (lachypjones) [reply cmd/workshop/exec.go:140]: Neither does shell. They'll be added to the environment inherited by the unit

Comment (lachypjones) [reply cmd/workshop/exec.go:140]: Also I believe the DBUS address is automatically inferred (as $XDG_RUNTIME_DIR/bus) so we should not need to set it explicitly.

File: cmd/workshop/exec.go (hunk 3 above)
Comment (akcano) [reply cmd/workshop/exec.go:97]: @lachypjones I'd say the exact full form of the corresponding `exec` command will do just fine.

**Comment by lachypjones** [comment]
Closing for now, to be addressed in conjunction with modifying interface environment variables.

**Pull request closed without merging**